task2/src/main/java/backend/Teacher.java

```diff
@@ -0,0 +1,120 @@
+package backend;
+
+
+import backend.layer.Layer;
+
+import java.lang.*;
+
+public class Teacher {
+    private NeuralNetwork neuralNetwork;
+    private double learingRate;
+    private double momentumRate;
+    double[] totalNeuronOutputErrorDerivativesOfOutputLayer;
+
+    public Teacher(NeuralNetwork neuralNetwork, double learingRate, double momentumRate) {
+        this.neuralNetwork = neuralNetwork;
+        this.learingRate = learingRate;
+        this.momentumRate = momentumRate;
+    }
+
+    public void changeWeightWithBackpropagation(int numOfEras, double[][] inputs, double[][] expectedOutputs) {
+        this.neuralNetwork.switchToLearningTime();
+        for (int i = 0; i < numOfEras; i++) {
+            for (int j = 0; j < inputs.length; j++) {
+                try {
+                    backpropagationAlgorithm(inputs[j], expectedOutputs[j]);
+                } catch (IllegalAccessException e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+        this.neuralNetwork.switchToWorkingTime();
+    }
+
+    private void backpropagationAlgorithm(double[] inputs, double[] expectedOutputs) throws IllegalAccessException {
+        double[] finalOutputs = neuralNetwork.calculateOutput(inputs);
+        double[] outputs = neuralNetwork.getLayersResult()[neuralNetwork.getNumberOfHiddenLayers() + 1];
+
+        double totalError = calculateTotalError(outputs, expectedOutputs);
+
+        double[] weightError = calculateErrorInOutputLayer(outputs, expectedOutputs);
+
+        double[][] weightErrorHidden = calculateErrorInHiddenLayers(finalOutputs);
+
+        editWeights(weightError, neuralNetwork.getOutputLayer());
+        for (int i = 0; i < neuralNetwork.getNumberOfHiddenLayers(); i++) {
+            editWeights(weightErrorHidden[i], neuralNetwork.getHiddenLayers()[neuralNetwork.getHiddenLayers().length - 1 - i]);
+        }
+    }
+
+    private void editWeights(double[] weightsError, Layer layer) throws IllegalAccessException {
+        for (int i = 0; i < layer.getNeurons().length; i++) {
+            int numOfWeightInNeuron = layer.getNeuron(i).getWeights().length;
+            for (int j = 0; j < numOfWeightInNeuron; j++) {
+                layer.getNeuron(i).setWeight(j, layer.getNeuron(i).getWeights()[j] - ((weightsError[i * layer.getNeuron(i).getWeights().length + j] * learingRate) + (momentumRate * layer.getLastWeightChange()[j])));
+                layer.setLastWeightChange(j, (weightsError[i * layer.getNeuron(i).getWeights().length + j] * learingRate) + (momentumRate * layer.getLastWeightChange()[j]));
```

Comment:
You have extracted this (`neuralNetwork.getHiddenLayers().length`) into numOfWeightInNeuron variable, don't calculate it again and weight change is calculated twicely.
```suggestion
                double weightChange = weightsError[i * numOfWeightInNeuron + j] * learingRate + momentumRate * layer.getLastWeightChange()[j];
                layer.getNeuron(i).setWeight(j, layer.getNeuron(i).getWeights()[j] - weightChange);
                layer.setLastWeightChange(j, weightChange);
```

Reply:
I changed it to getNumberOfHiddenLayers so problem does not occur any more.

Reply:
Problem is still occur, this part `weightsError[i * layer.getNeuron(i).getWeights().length + j] * learingRate) + (momentumRate * layer.getLastWeightChange()[j])` repeats twicely and creates train in 2 lines